In [ ]:
# from sklearn.preprocessing import LabelEncoder


# categorical_cols = ['status', 'sex', 'orientation', 'body_type', 'diet', 'drinks', 'drugs',
#                     'education', 'ethnicity', 'job', 'location', 'smokes', 'religion', 'sign', 'pets']


# label_encoding_cols = ['education', 'ethnicity', 'job', 'location', 'sign']


# label_encoders = {}
# for col in label_encoding_cols:
#     le = LabelEncoder()
#     df[col] = le.fit_transform(df[col].astype(str))  # Convert to string to avoid NaN issues
#     label_encoders[col] = le  # Save the encoder in case you need to decode later


# df = pd.get_dummies(df, columns=[col for col in categorical_cols if col not in label_encoding_cols], drop_first=True)


# encoded_file_path = "okcupid_profiles_encoded.csv"  # Output file path
# df.to_csv(encoded_file_path, index=False)

# print(f"Dataset with encoding saved at: {encoded_file_path}")

# Amooora text data cleaning

## Import libraries and load csv

In [30]:
import pandas as pd
import numpy as np
import string
from nltk.tokenize import word_tokenize
from nltk.stem.wordnet import WordNetLemmatizer
from sklearn.feature_extraction.text import TfidfVectorizer
import nltk
from nltk.corpus import stopwords
# nltk.download('punkt_tab')

In [2]:
!tree ..

..
├── notebooks
│   └── dedemenezes_exploratory.ipynb
├── raw_data
│   ├── okcupid_profiles.csv
│   ├── okcupid_profiles_encoded.csv
│   ├── parsed_data_public.csv
│   ├── question_data.csv
│   └── user_data_public.csv
├── README.md
└── requirements.txt

2 directories, 8 files


In [3]:
df = pd.read_csv('../raw_data/okcupid_profiles.csv')

In [4]:
df.shape

(59946, 31)

In [5]:
df.head(2)

age  status sex orientation       body_type               diet    drinks  \
0   22  single   m    straight  a little extra  strictly anything  socially   
1   35  single   m    straight         average       mostly other     often   

       drugs                      education     ethnicity  ...  \
0      never  working on college/university  asian, white  ...   
1  sometimes          working on space camp         white  ...   

                                              essay0  \
0  about me:  i would love to think that i was so...   
1  i am a chef: this is what that means. 1. i am ...   

                                              essay1  \
0  currently working as an international agent fo...   
1  dedicating everyday to being an unbelievable b...   

                                              essay2  \
0  making people laugh. ranting about a good salt...   
1  being silly. having ridiculous amonts of fun w...   

                                              essay3  \
0  the way i look. i am a six foot half asian, ha...   
1                                                NaN   

                                              essay4  \
0  books: absurdistan, the republic, of mice and ...   
1  i am die hard christopher moore fan. i don't r...   

                                              essay5  \
0                  food. water. cell phone. shelter.   
1  delicious porkness in all of its glories. my b...   

                        essay6  \
0  duality and humorous things   
1                          NaN   

                                              essay7  \
0  trying to find someone to hang out with. i am ...   
1                                                NaN   

                                              essay8  \
0  i am new to california and looking for someone...   
1  i am very open and will share just about anyth...   

                                              essay9  
0  you want to be swept off your feet! you are ti...  
1                                                NaN  

[2 rows x 31 columns]

## Access Text Dataset

In [6]:
df.iloc[:, :21]

age     status sex orientation       body_type               diet  \
0       22     single   m    straight  a little extra  strictly anything   
1       35     single   m    straight         average       mostly other   
2       38  available   m    straight            thin           anything   
3       23     single   m    straight            thin         vegetarian   
4       29     single   m    straight        athletic                NaN   
...    ...        ...  ..         ...             ...                ...   
59941   59     single   f    straight             NaN                NaN   
59942   24     single   m    straight             fit    mostly anything   
59943   42     single   m    straight         average    mostly anything   
59944   27     single   m    straight        athletic    mostly anything   
59945   39     single   m         gay         average                NaN   

           drinks      drugs                          education  \
0        socially      never      working on college/university   
1           often  sometimes              working on space camp   
2        socially        NaN     graduated from masters program   
3        socially        NaN      working on college/university   
4        socially      never  graduated from college/university   
...           ...        ...                                ...   
59941    socially      never  graduated from college/university   
59942       often  sometimes      working on college/university   
59943  not at all      never     graduated from masters program   
59944    socially      often      working on college/university   
59945    socially        NaN     graduated from masters program   

                 ethnicity  ...  income                           job  \
0             asian, white  ...      -1                transportation   
1                    white  ...   80000          hospitality / travel   
2                      NaN  ...      -1                           NaN   
3                    white  ...   20000                       student   
4      asian, black, other  ...      -1   artistic / musical / writer   
...                    ...  ...     ...                           ...   
59941                  NaN  ...      -1   sales / marketing / biz dev   
59942         white, other  ...      -1         entertainment / media   
59943                asian  ...  100000  construction / craftsmanship   
59944         asian, black  ...      -1             medicine / health   
59945                white  ...      -1             medicine / health   

            last_online                         location  \
0      2012-06-28-20-30  south san francisco, california   
1      2012-06-29-21-41              oakland, california   
2      2012-06-27-09-10        san francisco, california   
3      2012-06-28-14-22             berkeley, california   
4      2012-06-27-21-26        san francisco, california   
...                 ...                              ...   
59941  2012-06-12-21-47              oakland, california   
59942  2012-06-29-11-01        san francisco, california   
59943  2012-06-27-23-37  south san francisco, california   
59944  2012-06-23-13-01        san francisco, california   
59945  2012-06-29-00-42        san francisco, california   

                                    offspring                       pets  \
0      doesn't have kids, but might want them  likes dogs and likes cats   
1      doesn't have kids, but might want them  likes dogs and likes cats   
2                                         NaN                   has cats   
3                           doesn't want kids                 likes cats   
4                                         NaN  likes dogs and likes cats   
...                                       ...                        ...   
59941                                has kids                   has dogs   
59942                       doesn't have kids  likes dogs and likes cats   
59943  

In [7]:
text_df = df.iloc[:, 21:]

In [8]:
text_df.head()

essay0  \
0  about me:  i would love to think that i was so...   
1  i am a chef: this is what that means. 1. i am ...   
2  i'm not ashamed of much, but writing public te...   
3          i work in a library and go to school. . .   
4  hey how's it going? currently vague on the pro...   

                                              essay1  \
0  currently working as an international agent fo...   
1  dedicating everyday to being an unbelievable b...   
2  i make nerdy software for musicians, artists, ...   
3          reading things written by old dead people   
4                         work work work work + play   

                                              essay2  \
0  making people laugh. ranting about a good salt...   
1  being silly. having ridiculous amonts of fun w...   
2  improvising in different contexts. alternating...   
3  playing synthesizers and organizing books acco...   
4  creating imagery to look at: http://bagsbrown....   

                                              essay3  \
0  the way i look. i am a six foot half asian, ha...   
1                                                NaN   
2  my large jaw and large glasses are the physica...   
3                  socially awkward but i do my best   
4            i smile a lot and my inquisitive nature   

                                              essay4  \
0  books: absurdistan, the republic, of mice and ...   
1  i am die hard christopher moore fan. i don't r...   
2  okay this is where the cultural matrix gets so...   
3  bataille, celine, beckett. . . lynch, jarmusch...   
4  music: bands, rappers, musicians at the moment...   

                                              essay5  \
0                  food. water. cell phone. shelter.   
1  delicious porkness in all of its glories. my b...   
2  movement conversation creation contemplation t...   
3                                                NaN   
4                                                NaN   

                        essay6  \
0  duality and humorous things   
1                          NaN   
2                          NaN   
3   cats and german philosophy   
4                          NaN   

                                              essay7  \
0  trying to find someone to hang out with. i am ...   
1                                                NaN   
2  viewing. listening. dancing. talking. drinking...   
3                                                NaN   
4                                                NaN   

                                              essay8  \
0  i am new to california and looking for someone...   
1  i am very open and will share just about anyth...   
2  when i was five years old, i was known as "the...   
3                                                NaN   
4                                                NaN   

                                              essay9  
0  you want to be swept off your feet! you are ti...  
1                                                NaN  
2  you are bright, open, intense, silly, ironic, ...  
3                              you feel so inclined.  
4                                                NaN

## Checking dataset

In [9]:
text_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59946 entries, 0 to 59945
Data columns (total 10 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   essay0  54458 non-null  object
 1   essay1  52374 non-null  object
 2   essay2  50308 non-null  object
 3   essay3  48470 non-null  object
 4   essay4  49409 non-null  object
 5   essay5  49096 non-null  object
 6   essay6  46175 non-null  object
 7   essay7  47495 non-null  object
 8   essay8  40721 non-null  object
 9   essay9  47343 non-null  object
dtypes: object(10)
memory usage: 4.6+ MB


### Duplicates

In [10]:
text_df.duplicated().sum()

np.int64(2134)

In [11]:
text_df[text_df.duplicated()]

essay0 essay1 essay2 essay3 essay4 essay5 essay6 essay7 essay8 essay9
51       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
64       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
80       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
84       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
175      NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...
59757    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
59791    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
59857    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
59881    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
59930    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN

[2134 rows x 10 columns]

We have `2134`duplicated observations all made of empty answers to all questions

In [12]:
# no_dup_df = text_df.drop_duplicates()
# no_dup_df.shape

### Empty values

In [13]:
text_df.isnull().sum().sort_values(ascending=False) / len(text_df) * 100

essay8    32.070530
essay6    22.972342
essay9    21.023922
essay7    20.770360
essay3    19.143896
essay5    18.099623
essay4    17.577486
essay2    16.077803
essay1    12.631368
essay0     9.154906
dtype: float64

In [14]:
text_df.dropna().shape

(29866, 10)

In [15]:
text_df.shape

(59946, 10)

Drop is not an option!

## % of answered questions

In [16]:
text_df.columns

Index(['essay0', 'essay1', 'essay2', 'essay3', 'essay4', 'essay5', 'essay6',
       'essay7', 'essay8', 'essay9'],
      dtype='object')

In [17]:
text_df['not_answered_percent'] = text_df.apply(lambda x: x.isna().sum() / len(text_df.columns) * 100, axis=1)

In [18]:
text_df

essay0  \
0      about me:  i would love to think that i was so...   
1      i am a chef: this is what that means. 1. i am ...   
2      i'm not ashamed of much, but writing public te...   
3              i work in a library and go to school. . .   
4      hey how's it going? currently vague on the pro...   
...                                                  ...   
59941  vibrant, expressive, caring optimist. i love b...   
59942  i'm nick. i never know what to write about mys...   
59943  hello! i enjoy traveling, watching movies, and...   
59944  "all i have in this world are my balls and my ...   
59945  is it odd that having a little "enemy" status ...   

                                                  essay1  \
0      currently working as an international agent fo...   
1      dedicating everyday to being an unbelievable b...   
2      i make nerdy software for musicians, artists, ...   
3              reading things written by old dead people   
4                             work work work work + play   
...                                                  ...   
59941  the happiest times have been when life came to...   
59942  currently finishing school for film production...   
59943  i'm a civil engineer, who enjoys helping the c...   
59944  following my dreams... "you got a dream... you...   
59945  i work with elderly people (psychotherapy and ...   

                                                  essay2  \
0      making people laugh. ranting about a good salt...   
1      being silly. having ridiculous amonts of fun w...   
2      improvising in different contexts. alternating...   
3      playing synthesizers and organizing books acco...   
4      creating imagery to look at: http://bagsbrown....   
...                                                  ...   
59941  i make an outstanding osso bucco. i am also ve...   
59942  filmmaking, photography, graphic design, web d...   
59943  - looking at things objectively - getting thin...   
59944                                          listening   
59945  i'm a great bullshitter. i don't know what it ...   

                                                  essay3  \
0      the way i look. i am a six foot half asian, ha...   
1                                                    NaN   
2      my large jaw and large glasses are the physica...   
3                      socially awkward but i do my best   
4                i smile a lot and my inquisitive nature   
...                                                  ...   
59941  i am told that people notice my smile, eyes an...   
59942                                dude, i don't know.   
59943  i'm quiet until i get used to the environment ...   
59944  it used to be the hair until i mowed it off bu...   
59945  either that i am funny/sarcastic, or that i am...   

                                                  essay4  \
0      books: absurdistan, the republic, of mice and ...   
1      i am die hard christopher moore fan. i don't r...   
2      okay this is where the cultural matrix gets so...   
3      bataille, celine, beckett. . . lynch, jarmusch...   
4      music: bands, rappers, musicians at the moment...   
...                                                  ...   
59941  i am an avid movie watcher and follow the broa...   
59942  movies: hook (the greatest adventure ever!), g...   
59943  last book: "game change". movies: bourne serie...   
59944  where to begin musically: right now i listen t...   
59945  i just read the help by kathryn stockett, sooo...   

                                                  essay5  \
0                      food. water. cell phone. shelter.   
1      delicious porkness in all of its glories. my b...   
2      movement conversation creation contemplation t...   
3                                                    NaN   
4                                                    NaN   
...                                                  ...   
59941         my family, 

## Replace NaN with empty string

In [19]:
no_nan_df = text_df.replace(np.nan, '')

In [20]:
no_nan_df.isna()

essay0  essay1  essay2  essay3  essay4  essay5  essay6  essay7  essay8  \
0       False   False   False   False   False   False   False   False   False   
1       False   False   False   False   False   False   False   False   False   
2       False   False   False   False   False   False   False   False   False   
3       False   False   False   False   False   False   False   False   False   
4       False   False   False   False   False   False   False   False   False   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
59941   False   False   False   False   False   False   False   False   False   
59942   False   False   False   False   False   False   False   False   False   
59943   False   False   False   False   False   False   False   False   False   
59944   False   False   False   False   False   False   False   False   False   
59945   False   False   False   False   False   False   False   False   False   

       essay9  not_answered_percent  
0       False                 False  
1       False                 False  
2       False                 False  
3       False                 False  
4       False                 False  
...       ...                   ...  
59941   False                 False  
59942   False                 False  
59943   False                 False  
59944   False                 False  
59945   False                 False  

[59946 rows x 11 columns]

## Combining into one big text column

In [21]:
no_nan_df['combined'] = no_nan_df.apply(lambda row: ' '.join(row.values.astype(str)), axis=1)

In [22]:
no_nan_df.combined

0        about me:  i would love to think that i was so...
1        i am a chef: this is what that means. 1. i am ...
2        i'm not ashamed of much, but writing public te...
3        i work in a library and go to school. . . read...
4        hey how's it going? currently vague on the pro...
                               ...                        
59941    vibrant, expressive, caring optimist. i love b...
59942    i'm nick. i never know what to write about mys...
59943    hello! i enjoy traveling, watching movies, and...
59944    "all i have in this world are my balls and my ...
59945    is it odd that having a little "enemy" status ...
Name: combined, Length: 59946, dtype: object

## Cleaning text

In [23]:
def basic_cleaning(sentence):
    # Basic cleaning
    sentence = sentence.strip() # remove whitespaces
    sentence = sentence.lower() # lowercase
    sentence = ''.join(char for char in sentence if not char.isdigit()) # remove numbers
    
    # Advanced cleaning
    for punctuation in string.punctuation:
        sentence = sentence.replace(punctuation, '')
        
    return sentence

In [24]:
cleaned_df = pd.DataFrame()

# Criar Novas colunas essayX_clean
essays = no_nan_df.columns[:10]
for essay in essays:
    cleaned_df[f"{essay}_clean"] = no_nan_df[essay].apply(basic_cleaning)

cleaned_df['combined_clean'] = no_nan_df.combined.apply(basic_cleaning)

cleaned_df.columns

Index(['essay0_clean', 'essay1_clean', 'essay2_clean', 'essay3_clean',
       'essay4_clean', 'essay5_clean', 'essay6_clean', 'essay7_clean',
       'essay8_clean', 'essay9_clean', 'combined_clean'],
      dtype='object')

## Preprocessing

### Tokenization

Let's take one row and play with **Tokenization**, **stopwords** and **lemmatize** 

In [29]:
text = no_nan_df.loc[0, 'essay0']

word_tokens = word_tokenize(text)

pd.DataFrame({'essay0_words': word_tokens})

essay0_words
0          about
1             me
2              :
3              i
4          would
..           ...
264            a
265         good
266    attention
267         span
268            .

[269 rows x 1 columns]

### Remove stopwords

In [33]:
word_tokens

stop_words = set(stopwords.words('english'))

pd.DataFrame()

stopwords_removed = [w for w in word_tokens if w in stop_words]

tokens_cleaned = [w for w in word_tokens if not w in stop_words]
tokens_cleaned

[':',
 'would',
 'love',
 'think',
 'kind',
 'intellectual',
 ':',
 'either',
 'dumbest',
 'smart',
 'guy',
 ',',
 'smartest',
 'dumb',
 'guy',
 '.',
 'ca',
 "n't",
 'say',
 'tell',
 'difference',
 '.',
 'love',
 'talk',
 'ideas',
 'concepts',
 '.',
 'forge',
 'odd',
 'metaphors',
 'instead',
 'reciting',
 'cliches',
 '.',
 'like',
 'simularities',
 'friend',
 'mine',
 "'s",
 'house',
 'underwater',
 'salt',
 'mine',
 '.',
 'favorite',
 'word',
 'salt',
 'way',
 '(',
 'weird',
 'choice',
 'know',
 ')',
 '.',
 'things',
 'life',
 'better',
 'metaphors',
 '.',
 'seek',
 'make',
 'little',
 'better',
 'everyday',
 ',',
 'productively',
 'lazy',
 'way',
 '.',
 'got',
 'tired',
 'tying',
 'shoes',
 '.',
 'considered',
 'hiring',
 'five',
 'year',
 'old',
 ',',
 'would',
 'probably',
 'tie',
 'shoes',
 '...',
 'decided',
 'wear',
 'leather',
 'shoes',
 'dress',
 'shoes',
 '.',
 ':',
 'love',
 'really',
 'serious',
 ',',
 'really',
 'deep',
 'conversations',
 'really',
 'silly',
 'stuff',
 '.

### Lemmatizing

In [34]:
# Lemmatizing the verbs
verb_lemmatized = [
    WordNetLemmatizer().lemmatize(word, pos = "v") # v --> verbs
    for word in tokens_cleaned
]

# 2 - Lemmatizing the nouns
noun_lemmatized = [
    WordNetLemmatizer().lemmatize(word, pos = "n") # n --> nouns
    for word in verb_lemmatized
]

pd.DataFrame(dict(token=tokens_cleaned, verbs=verb_lemmatized, noun=noun_lemmatized))

### All together

In [38]:
def basic_preprocessing(sentence):
    tokenized_sentence = word_tokenize(sentence) ## tokenize
    stop_words = set(stopwords.words('english')) ## define stopwords

    tokenized_sentence_cleaned = [ ## remove stopwords
        w for w in tokenized_sentence if not w in stop_words
    ]

    lemmatized = [
        WordNetLemmatizer().lemmatize(word, pos = "v")
        for word in tokenized_sentence_cleaned
    ]

    # 2 - Lemmatizing the nouns
    noun_lemmatized = [
        WordNetLemmatizer().lemmatize(word, pos = "n")
        for word in lemmatized
    ]
    
    cleaned_sentence = ' '.join(word for word in noun_lemmatized)
    return cleaned_sentence

text = no_nan_df.loc[0, 'essay0']
print(f"Original text:\n{text}")
basic_preprocessing(text)

Original text:
about me:  i would love to think that i was some some kind of intellectual: either the dumbest smart guy, or the smartest dumb guy. can't say i can tell the difference. i love to talk about ideas and concepts. i forge odd metaphors instead of reciting cliches. like the simularities between a friend of mine's house and an underwater salt mine. my favorite word is salt by the way (weird choice i know). to me most things in life are better as metaphors. i seek to make myself a little better everyday, in some productively lazy way. got tired of tying my shoes. considered hiring a five year old, but would probably have to tie both of our shoes... decided to only wear leather shoes dress shoes.  about you:  you love to have really serious, really deep conversations about really silly stuff. you have to be willing to snap me out of a light hearted rant with a kiss. you don't have to be funny, but you have to be able to make me laugh. you should be able to bend spoons with your 

": would love think kind intellectual : either dumbest smart guy , smartest dumb guy . ca n't say tell difference . love talk idea concept . forge odd metaphor instead recite cliche . like simularities friend mine 's house underwater salt mine . favorite word salt way ( weird choice know ) . thing life better metaphor . seek make little better everyday , productively lazy way . get tire tie shoe . consider hire five year old , would probably tie shoe ... decide wear leather shoe dress shoe . : love really serious , really deep conversation really silly stuff . will snap light hearted rant kiss . n't funny , able make laugh . able bend spoon mind , telepathically make smile still work . love life , cool let wind blow . extra point read guess favorite video game ( hint give yet ) . lastly good attention span ."

In [39]:
cleaned_df.essay0_clean.apply(basic_preprocessing)

0        would love think kind intellectual either dumb...
1        chef mean workaholic love cook regardless whet...
2        im ashamed much write public text online date ...
3                                   work library go school
4        hey hows go currently vague profile know come ...
                               ...                        
59941    vibrant expressive care optimist love people t...
59942    im nick never know write im sure hand im south...
59943    hello enjoy travel watch movie hang friend rul...
59944    world ball integrity one take either away momm...
59945    odd little enemy status someone make seem inte...
Name: essay0_clean, Length: 59946, dtype: object

In [ ]:
clean_preprocessed_df = cleaned_df.apply(lambda row: row.apply(basic_preprocessing))

In [ ]:
preprocessed_combined = no_nan_df.combined.apply(basic_preprocessing)

In [ ]:
preprocessed_combined.to_frame().rename(columns={'combined': 'combined_preprocess'})

In [ ]:
no_nan_df['combined_preprocess'] = preprocessed_combined

In [ ]:
no_nan_df.head(1)

## Vectorizing

Using relative frequency rather than count is robust to document length

Takes into account the context of the whole corpus

In [ ]:
# Instantiating the TfidfVectorizer
tf_idf_vectorizer = TfidfVectorizer()

# Training it on the texts
weighted_words = pd.DataFrame(tf_idf_vectorizer.fit_transform(no_nan_df.combined_preprocess).toarray(),
                 columns = tf_idf_vectorizer.get_feature_names_out())

weighted_words